File: hearts/src/GameEvent.hs
{-# LANGUAGE InstanceSigs #-}
module GameEvent where

import           Cards

import           Data.Map (Map)
import qualified Data.Map as Map

-- Event:
-- Objekt, das ein Ereignis beschreibt
-- - in der Vergangenheit
-- - fachlich
-- - sollten nicht den neuen Zustand enthalten
-- Event-Sourcing:
-- - Events erzählen die gesamte Geschichte der Domäne
-- - Redundanz OK

{-
data GameEvent =
    GameSetup [(Player, Hand)]
  | GameStarted Player
  | PlayedCard Player Card
  | TrickEnded [Card] Player
  | ReceivedPoints Player Integer
  | TurnChanged Player -- wer ist als nächstes dran
  | GameEnded Player
-}

data GameEvent
  = HandDealt Player Hand
  | PlayerTurnChanged Player
  | LegalCardPlayed Player Card
  | TrickTaken Player Trick
  | IllegalCardAttempted Player Card
  | GameEnded Player
  deriving (Show, Eq)

-- Command:
-- Objekt, der einen Wunsch repräsentiert
-- - in der Zukunft
-- - vielleicht passiert es gar nicht
data GameCommand =
    DealHands (Map Player Hand)
  | PlayCard Player Card
  deriving (Show, Eq)


data Game a = 
    RecordEvent GameEvent (() -> Game a)
  | IsPlayValid Player Card (Bool -> Game a)
  | TakeTrick (Maybe (Trick, Player) -> Game a)
  | NextPlayer Player (Player -> Game a)
  | GameOver (Maybe Player -> Game a)
  | Return a


instance Functor Game where
  -- fmap :: (a -> b) -> Game a -> Game b
  -- fmap = undefined

instance Applicative Game where
  -- pure :: a -> Game a
  -- pure = Return
  -- (<*>) :: Game (a -> b) -> Game a -> Game b
  -- (<*>) = undefined

instance Monad Game where
  return :: a -> Game a
  return = Return
  (>>=) :: Game a -> (a -> Game b) -> Game b
  (>>=) (RecordEvent event callback) next = RecordEvent event (\() -> callback () >>= next)
  (>>=) (Return result) next = next result
  (>>=) (IsPlayValid player card callback) next = IsPlayValid player card (\valid -> callback valid >>= next)
  (>>=) (TakeTrick callback) next = TakeTrick (\x -> callback x >>= next)
  (>>=) (NextPlayer player callback) next = NextPlayer player (\x -> callback x >>= next)
  (>>=) (GameOver callback) next = GameOver (\x -> callback x >>= next)

recordEventM :: GameEvent -> Game ()
recordEventM event = RecordEvent event Return

isPlayValidM :: Player -> Card -> Game Bool
isPlayValidM player card = IsPlayValid player card Return

takeTrickM :: Game (Maybe (Trick, Player))
takeTrickM = TakeTrick Return

nextPlayerM :: Player -> Game Player
nextPlayerM player = NextPlayer player Return

gameOverM :: Game (Maybe Player)
gameOverM = GameOver Return

-- Maybe Player: wer hat gewonnen, falls das Spiel vorbei ist
tableProcessCommandM :: GameCommand -> Game (Maybe Player)
tableProcessCommandM (DealHands hands) = do
  let events = map (uncurry HandDealt) (Map.toList hands)
  mapM_ recordEventM events
  return Nothing
tableProcessCommandM (PlayCard player card) = do
  valid <- isPlayValidM player card
  if valid
  then do
    recordEventM (LegalCardPlayed player card)
    trick <- takeTrickM
    case trick of
      Just (trick, trickTaker) -> do
        recordEventM (TrickTaken trickTaker trick)
        gameOver <- gameOverM
        case gameOver of
          Just winner -> do
            recordEventM (GameEnded winner)
            return (Just winner)
          Nothing -> do
            recordEventM (PlayerTurnChanged trickTaker)
            return Nothing
      Nothing -> do
        nextPlayer <- nextPlayerM player
        recordEventM (PlayerTurnChanged player)
        return Nothing
  else do
    recordEventM (IllegalCardAttempted player card)
    return Nothing

File: hearts-typescript-react/src/GameEvent.ts
export { GameEvent, GameCommand };

import { Player, Card, Hand, Trick } from './Cards.ts';

import { Map, List } from 'immutable';

type GameEvent =
    | { tag: 'HandDealt', player: Player, hand: Hand }
    | { tag: 'PlayerTurnChanged', player: Player }
    | { tag: 'LegalCardPlayed', player: Player, card: Card }
    | { tag: 'TrickTaken', player: Player, trick: Trick }
    | { tag: 'IllegalCardAttempted', player: Player, card: Card }
    | { tag: 'GameEnded', player: Player }

interface GameEventTypeRef {
    handDealt(player: Player, hand: Hand): GameEvent;
    playerTurnChanged(player: Player): GameEvent;
    legalCardPlayed(player: Player, card: Card): GameEvent;
    trickTaken(player: Player, trick: Trick): GameEvent;
    illegalCardAttempted(player: Player, card: Card): GameEvent;
    gameEnded(player: Player): GameEvent;

    caseOf<A>(event: GameEvent,
	      table: { HandDealt: (player: Player, hand: Hand) => A,
		       PlayerTurnChanged: (player: Player) => A,
		       LegalCardPlayed: (player: Player, card: Card) => A,
		       TrickTaken: (player: Player, trick: Trick) => A,
		       IllegalCardAttempted: (player: Player, card: Card) => A,
		       GameEnded: (player: Player) => A
		     }): A;
    // for testing
    deckToHandDealt(deck: List<Card>, players: List<Player>): List<GameEvent>;
}

const GameEvent: GameEventTypeRef = {
    handDealt: (player: Player, hand: Hand) =>
	({ tag: 'HandDealt', player: player, hand: hand }),
    playerTurnChanged: (player: Player) =>
	({ tag: 'PlayerTurnChanged', player: player }),
    legalCardPlayed: (player: Player, card: Card) =>
	({ tag: 'LegalCardPlayed', player: player, card: card }),
    trickTaken: (player: Player, trick: Trick) =>
	({ tag: 'TrickTaken', player: player, trick: trick }),
    illegalCardAttempted: (player: Player, card: Card) =>
	({ tag: 'IllegalCardAttempted', player: player, card: card }),
    gameEnded: (player: Player) =>
	({ tag: 'GameEnded', player: player }),

    caseOf: (event, table) => {
	switch (event.tag) {
	    case 'HandDealt': return table.HandDealt(event.player, event.hand);
	    case 'PlayerTurnChanged': return table.PlayerTurnChanged(event.player);
	    case 'LegalCardPlayed': return table.LegalCardPlayed(event.player, event.card);
	    case 'TrickTaken': return table.TrickTaken(event.player, event.trick);
	    case 'IllegalCardAttempted': return table.IllegalCardAttempted(event.player, event.card);
	    case 'GameEnded': return table.GameEnded(event.player);
	}
    },

    deckToHandDealt:
    (deck: List<Card>, players: List<Player>) => {
	let handSize = Math.floor(deck.size / players.size);
	return List([...Array(players.size).keys()].map(index =>
	    ({tag: 'HandDealt',
	      player: players.get(index)!,
	      hand: Hand.make(deck.skip(handSize * index).take(handSize)) })))
    }

}

type GameCommand =
    | { tag: 'DealHands', playerHands: Map<Player, Hand> }
    | { tag: 'PlayCard', player: Player, card: Card }

interface GameCommandTypeRef {
    dealHands(playerHands: Map<Player, Hand>): GameCommand;
    playCard(player: Player, card: Card): GameCommand;
    deckToDealHands(deck: List<Card>, players: List<Player>): GameCommand;
    caseOf<A>(command: GameCommand,
	      table: { DealHands: (playerHands: Map<Player, Hand>) => A,
		       PlayCard: (player: Player, card: Card) => A
		     }): A;
}

const GameCommand: GameCommandTypeRef = {
    dealHands: (playerHands: Map<Player, Hand>) =>
	({ tag: 'DealHands', playerHands: playerHands }),
    playCard: (player: Player, card: Card) =>
	({ tag: 'PlayCard', player: player, card: card }),

    deckToDealHands:
    (deck: List<Card>, players: List<Player>) => {
	let handSize = Math.floor(deck.size / players.size);
	return { tag: 'DealHands',
		 playerHands:
		 Map([...Array(players.size).keys()].map(index =>
		     [(players.get(index))!, Hand.make(deck.skip(handSize * index).take(handSize))])) }
		     
    },

    caseOf: (command, table) => {
	switch (command.tag) {
	    case 'DealHands': return table.DealHands(command.playerHands);
	    case 'PlayCard': return table.PlayCard(command.player, command.card);
	}
    }
}

